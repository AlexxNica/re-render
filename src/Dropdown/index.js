import React from 'react';
import PropTypes from 'prop-types';
import { ListView, Text, TouchableOpacity, View } from 'react-native';
import { Icon } from 'react-native-elements';

const styles = {
    container: {
        width: 200,
    },
    autoCompleteInput: {
        width: 200,
        height: 40,
        justifyContent: 'center',
        borderWidth: 2,
        borderStyle: 'solid',
        borderColor: '#e0e0e0',
    },
    listContainer: {
        zIndex: 1,
        position: 'absolute',
        marginTop: 40,
        backgroundColor: '#FFF',
        width: 200
    },
    listItem: {
        height: 40,
        justifyContent: 'center',
        borderWidth: 2,
        borderStyle: 'solid',
        borderColor: '#e0e0e0',
        borderTopWidth: 0,
    },
};

export default class Dropdown extends React.Component {
    constructor(props) {
        super(props);

        const ds = new ListView.DataSource({rowHasChanged: (r1, r2) => r1 !== r2});
        this.state = {
            dataSource: ds.cloneWithRows(props.list),
            showList: false,
            selectedItem: ''
        };
    }

    componentWillReceiveProps(nextProps) {
        const ds = new ListView.DataSource({rowHasChanged: (r1, r2) => r1 !== r2});
        this.setState({
            dataSource: ds.cloneWithRows(nextProps.list),
        })
    }

    render() {
        return (
            <View style={styles.container}>
                <TouchableOpacity
                    style={[styles.autoCompleteInput, {backgroundColor: this.props.placeholderBackground}]}
                    onPress={() => {
                        this.setState({showList: !this.state.showList})
                    }}>
                    <View
                        style={[{flexDirection: 'row', marginRight: 10, marginLeft: 10}]}>
                        <Text style={{flex: 1, color: this.props.placeholderColor}}>
                            {this.state.selectedItem.value || this.props.placeholder}
                        </Text>
                        <Icon
                            name={this.state.showList ? 'angle-up' : 'angle-down'}
                            type='font-awesome'
                            color={this.props.placeholderColor || 'black'}
                        />
                    </View>
                </TouchableOpacity>

                {
                    this.state.showList
                    &&
                    <View style={styles.listContainer}>
                        <ListView
                            style={{maxHeight: this.props.listHeight}}
                            enableEmptySections={true}
                            dataSource={this.state.dataSource}
                            renderRow={
                                (rowData) => (
                                    <TouchableOpacity
                                        style={[styles.listItem, this.props.itemStyle]}
                                        onPress={() => {
                                            this.setState({showList: false, selectedItem: rowData});
                                            if (this.props.onItemSelected)
                                                this.props.onItemSelected(rowData);
                                        }}>
                                        <Text style={{marginLeft: 10}}>
                                            {rowData.value}
                                        </Text>
                                    </TouchableOpacity>
                                )
                            }
                        />
                    </View>
                }

            </View>
        )
    }
}

Dropdown.defaultProps = {
    listHeight: 200
};

Dropdown.propTypes = {
    /**
     * An array of objects of the format {key:1,value:''}
     */
    list: PropTypes.arrayOf(PropTypes.shape({
        key: PropTypes.number.isRequired,
        value: PropTypes.string.isRequired,
    })).isRequired,
    /**
     * The default text to be shown when nothing is selected.
     */
    placeholder: PropTypes.string,
    /**
     * Callback to be invoked when a dropdown item is selected.
     */
    onItemSelected: PropTypes.func,
    /**
     * Height of the dropdown list(default height = 200)
     */
    listHeight: PropTypes.number,
    /**
     * Background color of dropdown's in collapsed state.
     */
    placeholderBackground: PropTypes.string,
    /**
     * Text and icon color of dropdown in collapsed state.
     */
    placeholderColor: PropTypes.string,
    /**
     * Style of single list item.
     */
    itemStyle: PropTypes.object,
};
