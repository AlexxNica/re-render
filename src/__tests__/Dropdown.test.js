import React from 'react';
import renderer from 'react-test-renderer';
import Dropdown from '../Dropdown/index';

it('renders correctly', () => {
    const tree = renderer.create(
        <Dropdown list={[{key: 1, value: 'Apple'}, {key: 2, value: 'Mango'}]}
                  placeholder={'Choose a fruit'}
                  listHeight={100}
                  placeholderColor={'red'}
                  placeholderBackground={'blue'}
                  itemStyle={{backgroundColor: 'red'}}
                  onItemSelected={(item) => {
                  }}
        />
    ).toJSON();
    expect(tree).toMatchSnapshot();
});
