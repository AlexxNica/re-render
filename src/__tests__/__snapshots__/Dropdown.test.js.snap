// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="rn-alignItems-1oszu61 rn-borderTopStyle-1efd50x rn-borderRightStyle-14skgim rn-borderBottomStyle-rull8r rn-borderLeftStyle-mm0ijv rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-boxSizing-deolkf rn-display-6koalj rn-flexShrink-1qe8dj5 rn-flexBasis-1mlwlqe rn-flexDirection-eqz5dr rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-minHeight-ifefl9 rn-minWidth-bcqeeo rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j rn-paddingLeft-gy4na3 rn-position-bnwqim"
  style={
    Object {
      "width": "200px",
    }
  }
>
  <div
    className="rn-alignItems-1oszu61 rn-borderTopStyle-1efd50x rn-borderRightStyle-14skgim rn-borderBottomStyle-rull8r rn-borderLeftStyle-mm0ijv rn-borderTopWidth-1jxfwug rn-borderRightWidth-18p6if4 rn-borderBottomWidth-wgabs5 rn-borderLeftWidth-dwliz8 rn-boxSizing-deolkf rn-cursor-1loqt21 rn-display-6koalj rn-flexShrink-1qe8dj5 rn-flexBasis-1mlwlqe rn-flexDirection-eqz5dr rn-height-eu3ka rn-justifyContent-1777fci rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-minHeight-ifefl9 rn-minWidth-bcqeeo rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j rn-paddingLeft-gy4na3 rn-position-bnwqim rn-transitionDuration-eafdt9 rn-transitionProperty-1i6wzkk rn-userSelect-lrvibr"
    onKeyDown={[Function]}
    onKeyPress={[Function]}
    onKeyUp={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "backgroundColor": "blue",
        "borderBottomColor": "rgba(224,224,224,1)",
        "borderLeftColor": "rgba(224,224,224,1)",
        "borderRightColor": "rgba(224,224,224,1)",
        "borderTopColor": "rgba(224,224,224,1)",
        "width": "200px",
      }
    }
    tabIndex="0"
  >
    <div
      className="rn-alignItems-1oszu61 rn-borderTopStyle-1efd50x rn-borderRightStyle-14skgim rn-borderBottomStyle-rull8r rn-borderLeftStyle-mm0ijv rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-boxSizing-deolkf rn-display-6koalj rn-flexShrink-1qe8dj5 rn-flexBasis-1mlwlqe rn-flexDirection-18u37iz rn-marginTop-1mnahxq rn-marginBottom-p1pxzi rn-marginLeft-1n0xq6e rn-marginRight-zso239 rn-minHeight-ifefl9 rn-minWidth-bcqeeo rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j rn-paddingLeft-gy4na3 rn-position-bnwqim"
    >
      <div
        className="rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-display-1471scf rn-flexGrow-16y2uox rn-flexShrink-1wbh5a2 rn-flexBasis-1mlwlqe rn-font-1lw9tu2 rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j rn-paddingLeft-gy4na3 rn-textDecoration-bauka4 rn-whiteSpace-q42fyq rn-wordWrap-qvutc0"
        dir="auto"
        style={
          Object {
            "color": "red",
          }
        }
      >
        Choose a fruit
      </div>
      <div
        className="rn-alignItems-1awozwy rn-backgroundColor-wib322 rn-borderTopStyle-1efd50x rn-borderRightStyle-14skgim rn-borderBottomStyle-rull8r rn-borderLeftStyle-mm0ijv rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-boxSizing-deolkf rn-display-6koalj rn-flexShrink-1qe8dj5 rn-flexBasis-1mlwlqe rn-flexDirection-eqz5dr rn-justifyContent-1777fci rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-minHeight-ifefl9 rn-minWidth-bcqeeo rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j rn-paddingLeft-gy4na3 rn-position-bnwqim"
        onPress={undefined}
        underlayColor="white"
      >
        <div
          className="rn-backgroundColor-wib322 rn-borderTopWidth-13yce4e rn-borderRightWidth-fnigne rn-borderBottomWidth-ndvcnb rn-borderLeftWidth-gxnn5r rn-display-1471scf rn-font-1lw9tu2 rn-marginTop-1mnahxq rn-marginRight-61z16t rn-marginBottom-p1pxzi rn-marginLeft-11wrixw rn-paddingTop-wk8lta rn-paddingRight-9aemit rn-paddingBottom-1mdbw0j rn-paddingLeft-gy4na3 rn-textDecoration-bauka4 rn-whiteSpace-q42fyq rn-wordWrap-qvutc0"
          dir="auto"
          style={
            Object {
              "color": "red",
              "fontFamily": "FontAwesome",
              "fontSize": "24px",
              "fontStyle": "normal",
              "fontWeight": "normal",
            }
          }
        >
          ï„‡
        </div>
      </div>
    </div>
  </div>
</div>
`;
